### ISSUE TITLE
[Bug] `selection-info` segment does not update after movement commands

### ISSUE THREAD

**Issue by JordanAnthonyKing** [seagle0128/doom-modeline#470]
### Thank you for the bug report

- [X] I am using the latest version of `doom-mode` related packages.
- [X] I checked [FAQ](https://github.com/seagle0128/doom-modeline#faq).
- [x] You may also try reproduce the issue using clean environment and minimal configurations with
the command `emacs -Q`.

### Bug description

Using a custom modeline like:
```
  (doom-modeline-def-modeline 'my-simple-line
    '(bar modals matches buffer-info persp-name remote-host word-count parrot selection-info) ; <--
    '(objed-state misc-info debug repl lsp major-mode process vcs checker))
```
makes the selection-info in the modeline only display `1C`, and does not update to reflect the selection.

### Steps to reproduce

Create a custom modeline as above with `selection-info` but without `buffer-position`

### Expected behavior

Selection info should be accurate in the modeline regardless of if buffer-position is displayed or not

### OS

Linux

### Emacs Version

gccemacs-28

### Emacs Configurations

Doom

**Comment by seagle0128** [comment]
Can you please provide more details? Did you observe the selection-info was inaccurate? any screenshot or recording?
I couldn't reproduce this issue locally.

**Comment by JordanAnthonyKing** [comment]
I believe I may have conflated the issue, see this screenshot:
![image](https://user-images.githubusercontent.com/58787893/135534220-adbef9de-7093-499f-ba61-83b51ab0e385.png)
5 lines are highlighted, but the modeline only lists 4. It seems that the modeline updates when something triggers it to, not when the cursor moves. In this lisp file that seems to happen quite a lot, when the cursor passes a variable for instance, but in org-mode:
![image](https://user-images.githubusercontent.com/58787893/135534400-4123bda6-427a-4ad3-844e-37831eda5baa.png)
nothing triggers the update, so selection-info only reports 1L.

**Comment by seagle0128** [comment]
I tried to reproduce this issue although I am not using doom, but I failed. It works well as expected.
I suggest to use `emacs -Q` to reproduce. Thus will figure out whether your personal configurations impact.

![emacs](https://user-images.githubusercontent.com/140797/135571697-4d7ecad4-312a-433b-b704-13c95090587e.gif)

**Comment by JordanAnthonyKing** [comment]
The gif you provided clearly demonstrates the issue. When buffer-position is used the modeline updates after every motion, when it isn't the modeline doesn't update while dragging moving, but something causes it to (seemingly the `concat...` popup) once you finish moving. Try again in an org file or other text file where nothing prompts the update.

**Comment by seagle0128** [comment]
Unfortunately I don't understand what's issue here.

![emacs](https://user-images.githubusercontent.com/140797/135605902-e2ff18aa-6e2f-49fa-b9d0-711378f30951.gif)

**Comment by JordanAnthonyKing** [comment]
I misconstrued the issue when I raised this bug. I'll retitle it.

Please try in org again, but with lines of plain text, no source blocks or other things that evaluate. If I use a source block similar to yours then the selection-info updates when the cursor overlays the region and the expression is evaluated. If you only select plain text, then the selection info never updates from 1C or 1L.

Again, you can see this in the first video you posted, with buffer-info disabled the selection info does not update until a popup for the currently hovered symbol is triggered.

The issue therefore is that the selection-info segment does not update on motion commands, and instead updates when something else causes the modeline to update (or some similar scenario). This is why selection-info works flawlessly with buffer-position, as buffer-position updates on every motion.

Text only:
![image](https://user-images.githubusercontent.com/58787893/135608612-ba887074-5119-46b3-b1ed-61a960226f28.png)

Passing the code block:
![image](https://user-images.githubusercontent.com/58787893/135608672-202f8654-b611-4522-a256-1fdf5d930422.png)

**Comment by seagle0128** [comment]
Can you reproduce with `emacs -Q`?

**Comment by JordanAnthonyKing** [comment]
In vanilla emacs as default:
![image](https://user-images.githubusercontent.com/58787893/135624983-74048caf-7c1c-4f3f-b3e8-fcb5a30f1232.png)

Then with following:
```  
(doom-modeline-def-modeline 'my-simple-line
    '(bar modals matches buffer-info-simple remote-host word-count selection-info parrot)
    '(objed-state misc-info debug repl lsp major-mode process vcs checker))
  (doom-modeline-set-modeline 'my-simple-line 'default)
```
Gives:
![image](https://user-images.githubusercontent.com/58787893/135625464-c22ade6b-5b0c-4868-a143-7e9920a3d789.png)

Selection info stays on 0C after setting mark and moving the cursor. This behaviour is also found while dragging the mouse but the selection info is updated on releasing it, probably because as I said before, the update of selection-info gets triggered by things being evaluated. _Mark set_ appears in the minibuffer when the mouse is released, and this triggers the selection info to be updated. Selection info is not updated on cursor movements alone however.

**Comment by seagle0128** [comment]
Well, I know how to reproduce the issue. It seems only exist on Emacs 28. I tried with Emacs 27 and all work well.  I checked the codes again, and no issue was found by me. It seems a bug of Emacs28 itself. Can you double check with other versions, like 26 or 27? Thanks!

**Comment by JordanAnthonyKing** [comment]
I still see this issue in emacs 27.2 on windows:
![image](https://user-images.githubusercontent.com/58787893/135714586-c25f46c7-6dbb-4d2f-ba55-ce5d1bdbcc96.png)

My earlier screenshots were of emacs 28 on WSL.

**Comment by seagle0128** [comment]
Wired! Can you please try evaluating the snippet below?

```elisp
(doom-modeline-def-segment selection-info
  "Information about the current selection, such as how many characters and
lines are selected, or the NxM dimensions of a block selection."
  (when (and (or mark-active (and (bound-and-true-p evil-local-mode)
                                  (eq evil-state 'visual)))
             (doom-modeline--active))
    (cl-destructuring-bind (beg . end)
      (if (and (bound-and-true-p evil-local-mode) (eq evil-state 'visual))
          (cons evil-visual-beginning evil-visual-end)
        (cons (region-beginning) (region-end)))
      (propertize
       (let ((lines (count-lines beg (min end (point-max)))))
         (concat (doom-modeline-spc)
                 (cond ((or (bound-and-true-p rectangle-mark-mode)
                            (and (bound-and-true-p evil-visual-selection)
                                 (eq 'block evil-visual-selection)))
                        (let ((cols (abs (- (doom-modeline-column end)
                                            (doom-modeline-column beg)))))
                          (format "%dx%dB" lines cols)))
                       ((and (bound-and-true-p evil-visual-selection)
                             (eq evil-visual-selection 'line))
                        (format "%dL" lines))
                       ((> lines 1)
                        (format "%dC %dL" (- end beg) lines))
                       (t
                        (format "%dC" (- end beg))))
                 (when doom-modeline-enable-word-count
                   (format " %dW" (count-words beg end)))
                 (doom-modeline-spc)))
       'face 'doom-modeline-highlight)
      (force-mode-line-update))))
```

**Comment by JordanAnthonyKing** [comment]
I get an warning evaling that statement and selection-info does not render:
![image](https://user-images.githubusercontent.com/58787893/135719562-5c6dc209-26f6-45af-b92a-280b0cef7b45.png)

**Comment by seagle0128** [comment]
Sorry, correct the snippet:
```elisp
(doom-modeline-def-segment selection-info
  "Information about the current selection, such as how many characters and
lines are selected, or the NxM dimensions of a block selection."
  (when (and (or mark-active (and (bound-and-true-p evil-local-mode)
                                  (eq evil-state 'visual)))
             (doom-modeline--active))
    (cl-destructuring-bind (beg . end)
      (if (and (bound-and-true-p evil-local-mode) (eq evil-state 'visual))
          (cons evil-visual-beginning evil-visual-end)
        (cons (region-beginning) (region-end)))
      (prog1
          (propertize
           (let ((lines (count-lines beg (min end (point-max)))))
             (concat (doom-modeline-spc)
                     (cond ((or (bound-and-true-p rectangle-mark-mode)
                                (and (bound-and-true-p evil-visual-selection)
                                     (eq 'block evil-visual-selection)))
                            (let ((cols (abs (- (doom-modeline-column end)
                                                (doom-modeline-column beg)))))
                              (format "%dx%dB" lines cols)))
                           ((and (bound-and-true-p evil-visual-selection)
                                 (eq evil-visual-selection 'line))
                            (format "%dL" lines))
                           ((> lines 1)
                            (format "%dC %dL" (- end beg) lines))
                           (t
                            (format "%dC" (- end beg))))
                     (when doom-modeline-enable-word-count
                       (format " %dW" (count-words beg end)))
                     (doom-modeline-spc)))
           'face 'doom-modeline-highlight)
        (force-mode-line-update)))))
```

**Comment by JordanAnthonyKing** [comment]
Same story:
![image](https://user-images.githubusercontent.com/58787893/135725938-cdb1b76c-897d-47de-acb7-d1c410f67b5e.png)

**Comment by seagle0128** [comment]
I did some troubleshooting today, and found the root cause is `region-begining` and `region-end` aren't updated timely in this case. No idea how to address it.

**Comment by JordanAnthonyKing** [comment]
Is there no hook we can add to update these variables?

**Comment by seagle0128** [comment]
`region-begining` and `region-end` are built-in C functions.

Try this snippet:

```elisp
(doom-modeline-def-segment selection-info
  (when (and (or mark-active (and (bound-and-true-p evil-local-mode)
                                  (eq evil-state 'visual)))
             (doom-modeline--active))
    (format " %d %d" (region-beginning) (region-end))))
```

**Comment by JordanAnthonyKing** [comment]
That outputs this:
![image](https://user-images.githubusercontent.com/58787893/136061176-79538050-a8e4-46bb-a0a5-23e7301b7e92.png)
But also doesn't update.

**Comment by seagle0128** [comment]
That demonstrates `(region-beginning)` and  `(region-end)` are not changed while moving the cursor in the region.

**Comment by JordanAnthonyKing** [comment]
¯\_(ツ)_/¯

**Comment by seagle0128** [comment]
@JordanAnthonyKing Did you upgraded to 28.2? Are you still seeing this issue? I tried with 28 and 29 recently and no issue there.

**Comment by JordanAnthonyKing** [comment]
I've kept the segment disabled in my modeline since :L I'll take a look tomorrow.

**Comment by SPFabGerman** [comment]
@seagle0128 I have encountered the same bug. I used the following configuration to reproduce it:
```emacs-lisp
(require 'package)

(setq debug-on-error t
      no-byte-compile t
      byte-compile-warnings nil
      inhibit-startup-screen t
      package-archives '(("melpa" . "https://melpa.org/packages/")
                         ("gnu" . "https://elpa.gnu.org/packages/"))
      package-user-dir (make-temp-file "min-init")
      custom-file (expand-file-name "custom.el" package-user-dir))

(delete-file package-user-dir)
(add-hook 'kill-emacs-hook `(lambda ()
                              (delete-directory ,package-user-dir t)))

(let* ((pkg-list '(evil doom-modeline)))

  (package-initialize)
  (package-refresh-contents)

  (mapc (lambda (pkg)
          (unless (package-installed-p pkg)
            (package-install pkg))
          (require pkg))
        pkg-list))

(doom-modeline-mode)
(evil-mode)
(line-number-mode -1)

(run-with-timer 0.1 0.1 (lambda () (when (region-active-p) (message "%s" (region-end)))))

```
I'm using Emacs 30.0.50 on Linux.

I could track the bug down to the `line-number-mode`. If I remove the line `(line-number-mode -1)` or enable `line-number-mode` after starting emacs the bug disappears and the `selection-info` is updated accordingly.

Using the timer above I could also see that `(region-end)` is updated accordingly (at least in the normal visual column mode), unlike earlier suspected. So I assume this is indeed an error with `doom-modeline` and not with emacs.

**Comment by seagle0128** [comment]
@SPFabGerman I believe this issue has addressed in f45a5a2.
